docs/operating-scylla/admin-tools/scylla-sstable.rst

```diff
@@ -34,19 +34,14 @@ You can specify more than one SStable.
 Schema
 ------
 
-All operations need a schema to interpret the SStables with.
-This tool tries to auto-detect the location of the ScyllaDB data directories and the name of the table the SStable belongs to.
-If the SStable is located in a ScyllaDB data directory, it works out-of-the-box, without any additional input from the user.
-If the SStable is located at an external path, you need to specify the names of the keyspace and table to which the SStable belongs. In addition, some hints as to where the ScyllaDB data directory is located may also be required.
+All operations need a schema to interpret the SStables with. By default the tool tries to auto-detect the schema of the SSTable. This should work out-of-the-box in most cases.
```

Comment:
Random idea: add a schema in the Scylla.db component (but translate every UDT to a compatible tuple)

Reply:
I just checked and the way we store the schema in the Statistics component already works with UDTs:
```
$ ./build/dev/scylla sstable --sstable-schema dump-data --logger-log-level scylla-sstable=trace /var/lib/scylla/data/ks/tbl-d6272950ac9511ef9a11892cdae71ae4/me-3gll_0mbr_2c8eo2ccgtk8wiqdwk-big-Data.db 1>out.txt
DEBUG 2024-11-27 04:00:07,811 [shard 0:main] scylla-sstable - Successfully read scylla.yaml from default location of conf/scylla.yaml
DEBUG 2024-11-27 04:00:07,811 [shard 0:main] scylla-sstable - Single schema source provided
DEBUG 2024-11-27 04:00:07,813 [shard 0:main] scylla-sstable - Succesfully loaded schema from  (/var/lib/scylla/data/ks/tbl-d6272950ac9511ef9a11892cdae71ae4/me-3gll_0mbr_2c8eo2ccgtk8wiqdwk-big-Data.db), obtained from --sstable-schema parameter
TRACE 2024-11-27 04:00:07,813 [shard 0:main] scylla-sstable - Loaded schema: org.apache.cassandra.config.CFMetaData@0x35b3b840[cfId=00179350-ac9e-11ef-a564-677590228781,ksName==ks,cfName=tbl,cfType=Standard,comparator=org.apache.cassandra.db.marshal.CompositeType(org.apache.cassandra.db.marshal.UTF8Type,org.apache.cassandra.db.marshal.ColumnToCollectionType(76:org.apache.cassandra.db.marshal.UserType(ks,7470,6631:org.apache.cassandra.db.marshal.Int32Type,6632:org.apache.cassandra.db.marshal.UTF8Type))),comment=,tombstoneGcOptions={"mode":"timeout","propagation_delay_in_seconds":"3600"},gcGraceSeconds=864000,minCompactionThreshold=4,maxCompactionThreshold=32,columnMetadata=[ColumnDefinition{name=$pk, type=org.apache.cassandra.db.marshal.Int32Type, kind=PARTITION_KEY, componentIndex=0, droppedAt=-9223372036854775808}, ColumnDefinition{name=v, type=org.apache.cassandra.db.marshal.UserType(ks,7470,6631:org.apache.cassandra.db.marshal.Int32Type,6632:org.apache.cassandra.db.marshal.UTF8Type), kind=REGULAR, componentIndex=null, droppedAt=-9223372036854775808}],compactionStrategyClass=class org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy,compactionStrategyOptions={enabled=true},compressionParameters={sstable_compression=org.apache.cassandra.io.compress.LZ4Compressor},bloomFilterFpChance=0.01,memtableFlushPeriod=0,caching={"keys":"ALL","rows_per_partition":"ALL"},cdc={},defaultTimeToLive=0,minIndexInterval=128,maxIndexInterval=2048,speculativeRetry=99.0PERCENTILE,triggers=[],isDense=false,version=00179351-ac9e-11ef-a564-677590228781,droppedColumns={},collections={76 : org.apache.cassandra.db.marshal.UserType(ks,7470,6631:org.apache.cassandra.db.marshal.Int32Type,6632:org.apache.cassandra.db.marshal.UTF8Type)},indices={}]
TRACE 2024-11-27 04:00:07,814 [shard 0:main] scylla-sstable - consume_sstables(): 1 sstables, merge=false, use_full_scan_reader=true
TRACE 2024-11-27 04:00:07,814 [shard 0:main] scylla-sstable - consume partition start
TRACE 2024-11-27 04:00:07,814 [shard 0:main] scylla-sstable - consume clustering row
TRACE 2024-11-27 04:00:07,814 [shard 0:main] scylla-sstable - consume partition end
TRACE 2024-11-27 04:00:07,814 [shard 0:main] scylla-sstable - consumer paused, skip_partition=false
```
This is how the column def is stored: `ColumnDefinition{name=v, type=org.apache.cassandra.db.marshal.UserType(ks,7470,6631:org.apache.cassandra.db.marshal.Int32Type,6632:org.apache.cassandra.db.marshal.UTF8Type), kind=REGULAR, componentIndex=null, droppedAt=-9223372036854775808}`

The schema:
```cql
cqlsh> create KEYSPACE ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};                                                                                                                                                             
                                                                                                                                                                                                                                                               
Warnings :                                                                                                                                                                                                                                                     
SimpleStrategy replication class is not recommended, but was used for keyspace ks. You may suppress this warning by delisting SimpleStrategy from replication_strategy_warn_list configuration option, or make it into an error by listing this replication str
ategy on replication_strategy_fail_list.                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                               
Using Replication Factor replication_factor=1 lower than the minimum_replication_factor_warn_threshold=3 is not recommended.                                                                                                                                   
                                                                                                                                                                                                                                                               
cqlsh> create TYPE ks.tp ( f1 int, f2 text) ;                                                                                                                                                                                                                  
cqlsh> create TABLE ks.tbl ( pk int PRIMARY KEY, v tp);  
```

Reply:
We just need to store the names of the key columns and the schema options we have a complete schema in the sstable. We can store these extra stuff in ScyllaMetadata component.